### PR TITLE
Add support for `SemiStructured` relational type

### PR DESCRIPTION
**Pull request by gayathrir11**
## How did you test this change?

- [x] Test(s) added
- [ ] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

File: .changeset/cool-poets-provide.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+"@finos/legend-extension-dsl-serializer": patch
```
Comment (akphi) [inline .changeset/cool-poets-provide.md:2]: why are we touching `DSL serializer`, this is relational-related so I expect to only see changes in `@finos/legend-graph`

File: packages/legend-extension-dsl-serializer/src/models/__tests__/DSLSerializer_Roundtrip.test.ts [hunk 2]
```diff
@@ -32,3 +33,10 @@ test(unitTest('DSL Serializer import resolution roundtrip'), async () => {
     pluginManager,
   );
 });
+
+test(unitTest('SemiStructured relational type roundtrip'), async () => {
```
Comment (akphi) [inline packages/legend-extension-dsl-serializer/src/models/__tests__/DSLSerializer_Roundtrip.test.ts:37]: Let's move this where it belongs

File: packages/legend-graph/src/models/metamodels/pure/packageableElements/store/relational/mapping/BindingTransformer.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { type Hashable, hashArray } from '@finos/legend-shared';
+import { action, computed, makeObservable, observable } from 'mobx';
+import { CORE_HASH_STRUCTURE } from '../../../../../../../MetaModelConst';
+
+export class BindingTransformer implements Hashable {
```
Comment (akphi) [inline packages/legend-graph/src/models/metamodels/pure/packageableElements/store/relational/mapping/BindingTransformer.ts:21]: Let's move this class to another directory
`/metamodels/pure/packageableElements/store/binding`
Also, we should move `Binding` to `@finos/legend-graph` now.

File: packages/legend-graph/src/models/protocols/pure/v1/model/packageableElements/store/relational/mapping/V1_BindingTransformer.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { hashArray, type Hashable } from '@finos/legend-shared';
+import { CORE_HASH_STRUCTURE } from '../../../../../../../../../MetaModelConst';
+
+export class V1_BindingTransformer implements Hashable {
```
Comment (akphi) [inline packages/legend-graph/src/models/protocols/pure/v1/model/packageableElements/store/relational/mapping/V1_BindingTransformer.ts:20]: Let's move this to the coressponding place `model/packageableElements/store/binding`

File: packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/V1_ProtocolToMetaModelPropertyMappingBuilder.ts [hunk 5]
```diff
@@ -533,6 +534,11 @@ export class V1_ProtocolToMetaModelPropertyMappingBuilder
       sourceSetImplementation,
       targetSetImplementation,
     );
+    if (protocol.bindingTransformer?.binding !== undefined) {
```
Comment (akphi) [inline packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/V1_ProtocolToMetaModelPropertyMappingBuilder.ts:537]: no need `!== undefined`

File: packages/legend-graph/src/models/protocols/pure/v1/transformation/pureProtocol/serializationHelpers/V1_DatabaseSerializationHelper.ts [hunk 6]
```diff
@@ -207,6 +209,12 @@ const V1_varCharRelationalDataTypeModelSchema = createModelSchema(V1_VarChar, {
   _type: usingConstantValueSchema(V1_RelationalDataTypeType.VARCHAR),
   size: primitive(),
 });
+const V1_semiStructuredRelationalDataTypeModelSchema = createModelSchema(
```
Comment (akphi) [inline packages/legend-graph/src/models/protocols/pure/v1/transformation/pureProtocol/serializationHelpers/V1_DatabaseSerializationHelper.ts:212]: add a new line please

**Review by akphi — DISMISSED** [review]
The PR change itself is quite simple, but it unravels something much bigger which is the fact that we are not out-of-sync with `engine` in terms of how we structure our code, `Binding` must be in core since relational depends on it, so I'd say we take the time to do this right.

@MauricioUyaguari could you see my comments and see if you agree with that folder structure I propose, else feel free to shoot your idea, I'm not as knowledgeable as you and Gayathri in this.

File: packages/legend-graph/src/models/metamodels/pure/packageableElements/store/relational/mapping/BindingTransformer.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  binding!: string;
```
Comment (akphi) [inline packages/legend-graph/src/models/metamodels/pure/packageableElements/store/relational/mapping/BindingTransformer.ts:22]: I missed this one, this should be `PackageableElementReference<Binding>`

**Comment by akphi** [comment]
@gayathrir11 @MauricioUyaguari As it stands, I think we need to move `Binding` and all other `external format` stuff back into core, i.e. `@finos/legend-graph`

@gs-ssh16 this is a module dependency in `finos/legend-studio` codebase which @MauricioUyaguari is trying to address by end of next week. We will merge this PR then. Sorry for the inconvenience!

File: packages/legend-graph/src/models/metamodels/pure/packageableElements/externalFormat/store/BindingTransformer.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  setBinding(value: PackageableElementReference<Binding>): void {
+    this.binding = value;
+  }
+
+  get hashCode(): string {
+    return hashArray([
+      CORE_HASH_STRUCTURE.BINDING_TRANSFORMER,
+      this.binding.value.path,
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/metamodels/pure/packageableElements/externalFormat/store/BindingTransformer.ts:41]: should this instead be `this.binding.hashValue` ?

File: packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/from/V1_MappingTransformer.ts [hunk 9]
```diff
@@ -490,6 +491,11 @@ const transformRelationalPropertyMapping = (
   propertyMapping.target = transformPropertyMappingTarget(
     element.targetSetImplementation,
   );
+  if (element.bindingTransformer?.binding !== undefined) {
+    const bindingTransformer = new V1_BindingTransformer();
+    bindingTransformer.binding = element.bindingTransformer.binding.value.path;
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/from/V1_MappingTransformer.ts:496]: `binding.value.valueForSerialization`

**Pull request merged**